src/fact_emitter.rs

```diff
@@ -0,0 +1,575 @@
+                    self.emit_subset_facts(bb, idx, &lhs_ty, expr, facts);
+                }
+
+                Statement::Drop(expr) => {
+                    // Evaluate the `expr`
+                    self.emit_expr_facts(bb, idx, expr, facts);
+
+                    // TODO: is there something more to do because we're in a "drop" ?
+                }
+            }
+        }
+    }
+
+    fn emit_expr_facts(&self, bb: &BasicBlock, idx: usize, expr: &Expr, facts: &mut Facts) {
+        match expr {
+            Expr::Access { kind, place } => {
+                match kind {
+                    // Borrowing clears its origin: it's issuing a fresh origin of the same name
+                    AccessKind::Borrow(origin) | AccessKind::BorrowMut(origin) => {
+                        facts
+                            .clear_origin
+                            .push((origin.into(), node_at(&bb.name, idx)));
+
+                        if matches!(kind, AccessKind::BorrowMut(_)) {
+                            // A mutable borrow is considered a write to the place:
+                            //
+                            // 1) it accesses the origins in the type
+                            let (_, origins) = self.ty_and_origins_of_place(place);
+                            for origin in origins {
+                                facts
+                                    .access_origin
+                                    .push((origin.clone(), node_at(&bb.name, idx)));
+                            }
+
+                            // 2) and invalidates existing loans of that place
+                            //
+                            // TODO: handle assignments to fields and loans taken on subsets of
+                            // their paths. Until then: only support invalidations on assignments
+                            // to complete places.
+                            //
+                            // TODO: here as well, there is a question of: can the loans we're
+                            // invalidating, reach the current node ?
+                            //
+                            if let Some(loans) = self.loans.get(place) {
+                                for (origin, _) in loans {
+                                    facts
+                                        .invalidate_origin
+                                        .push((origin.clone(), node_at(&bb.name, idx)));
+                                }
+                            }
+                        }
+                    }
+
+                    AccessKind::Copy | AccessKind::Move => {
+                        // FIXME: currently function call parameters are not parsed without access
+                        // kinds, check if there's some special behaviour needed for copy/moves,
+                        // instead of just being "reads" (e.g. maybe moves also need clearing
+                        // or invalidations)
```

Comment:
Yes, I think the purpose of disambiguating between copies and moves is that moves invalidate all borrows of the moved place whereas copies don't.